projects/packages/publicize/src/class-social-admin-page.php

```diff
@@ -0,0 +1,139 @@
+<?php
+/**
+ * Social Admin Page class.
+ *
+ * @package automattic/jetpack-publicize
+ */
+
+namespace Automattic\Jetpack\Publicize;
+
+use Automattic\Jetpack\Admin_UI\Admin_Menu;
+use Automattic\Jetpack\Assets;
+
+/**
+ * The class to handle the Social Admin Page.
+ */
+class Social_Admin_Page {
+
+	/**
+	 * The instance of the class.
+	 *
+	 * @var Social_Admin_Page
+	 */
+	private static $instance;
+
+	/**
+	 * Initialize the class.
+	 *
+	 * @return Social_Admin_Page
+	 */
+	public static function init() {
+		if ( ! isset( self::$instance ) ) {
+			self::$instance = new self();
+		}
+
+		return self::$instance;
+	}
+
+	/**
+	 * The constructor.
+	 */
+	private function __construct() {
+		add_action( 'admin_menu', array( $this, 'add_menu' ) );
+
+		/**
+		 * Admin_Menu::add_menu uses 1000, so we use 2000
+		 * to ensure we remove the old Social menu item after it has been added.
+		 */
+		add_action( 'admin_menu', array( $this, 'remove_old_social_menu_item' ), 2000 );
+	}
+
+	/**
+	 * Remove the page added by old versions of Social.
+	 */
+	public function remove_old_social_menu_item() {
+
+		if ( defined( 'JETPACK_SOCIAL_PLUGIN_ROOT_FILE' ) ) {
+
+			$plugin_data = get_plugin_data( (string) constant( 'JETPACK_SOCIAL_PLUGIN_ROOT_FILE' ), false, false );
+
+			$plugin_version = $plugin_data['Version'];
+
+			// If it's the old social version, remove the submenu page.
+			// TODO Update the version and operator before next Social release.
+			if ( version_compare( $plugin_version, '6.1.0', '<' ) ) {
+				/**
+				 * `add_submenu_page` allows multiple submenus with the same slug,
+				 * but `remove_submenu_page` only removes the first one it finds with the given slug.
+				 *
+				 * We add the menu using `admin_menu` hook unlike the old Social plugin,
+				 * which used the `init` hook, which runs before `admin_menu`.
+				 * So, the old Social plugin's menu is before the new one in $submenu global.
+				 *
+				 * So, `remove_submenu_page` should remove the menu added by the old Social plugin.
+				 *
+				 * @see https://developer.wordpress.org/reference/functions/add_submenu_page
+				 * @see https://developer.wordpress.org/reference/functions/remove_submenu_page
+				 */
+				remove_submenu_page( 'jetpack', 'jetpack-social' );
```

Comment:
How inefficient do you think it would be to simply call this before we add our page? It will return `false` if the submenu page isn't found, and we wouldn't have to do the social version checks etc.

Reply:
You have to add your menu via `Admin_Menu::add_menu()` before [`admin_menu` action with `1000` priority](https://github.com/Automattic/jetpack/blob/df67099b888ba72a4558fee7fbdc2e86f05ef992/projects/packages/admin-ui/src/class-admin-menu.php#L41). So, by that time it's not guaranteed that the old menu is registered or not. This is why I added that comment `Admin_Menu::add_menu uses 1000`

Reply:
A better solution would be to have [`Admin_Menu`](https://github.com/Automattic/jetpack/blob/a898ff9b9c8f070802b72fe581a16397233c184c/projects/packages/admin-ui/src/class-admin-menu.php) provide a method like `remove_menu` apart from [`add_menu`](https://github.com/Automattic/jetpack/blob/a898ff9b9c8f070802b72fe581a16397233c184c/projects/packages/admin-ui/src/class-admin-menu.php#L164)

Reply:
I mean more that we don't bother getting the Social plugin version etc, and just call remove_submenu_page. As it's added in init, we could hook into `init` late and call `remove_submenu_page`, and let Admin_Menu do its thing later in the request.

Reply:
No, that's not how it works. [`Admin_Menu`](https://github.com/Automattic/jetpack/blob/a898ff9b9c8f070802b72fe581a16397233c184c/projects/packages/admin-ui/src/class-admin-menu.php) keeps a list of menus added and then registers them on `admin_menu` action. So the menu added via `Admin_Menu::add_menu` may not even be registered when you call `remove_submenu_page`

Reply:
Oh annoying! In which case, let's stick with this and we can look at changing Admin_Menu in another PR